File: code-export-fixed/php/models/Course.php
<?php
/**
 * Course Model
 * نموذج الدورة
 */

require_once __DIR__ . '/../config/database.php';

class Course {
    private $conn;
    private $table = 'courses';

    public $id;
    public $title;
    public $description;
    public $instructor_id;
    public $category_id;
    public $thumbnail;
    public $price;
    public $discount_price;
    public $level;
    public $duration;
    public $is_featured;
    public $is_published;
    public $created_at;
    public $updated_at;

    public function __construct() {
        $database = new Database();
        $this->conn = $database->getConnection();
    }

    public function create() {
        $query = "INSERT INTO " . $this->table . " 
                  (title, description, instructor_id, category_id, thumbnail, price, discount_price, level, duration, is_featured, is_published) 
                  VALUES (:title, :description, :instructor_id, :category_id, :thumbnail, :price, :discount_price, :level, :duration, :is_featured, :is_published)";

        $stmt = $this->conn->prepare($query);

        $stmt->bindParam(':title', $this->title);
        $stmt->bindParam(':description', $this->description);
        $stmt->bindParam(':instructor_id', $this->instructor_id);
        $stmt->bindParam(':category_id', $this->category_id);
        $stmt->bindParam(':thumbnail', $this->thumbnail);
        $stmt->bindParam(':price', $this->price);
        $stmt->bindParam(':discount_price', $this->discount_price);
        $stmt->bindParam(':level', $this->level);
        $stmt->bindParam(':duration', $this->duration);
        $stmt->bindParam(':is_featured', $this->is_featured);
        $stmt->bindParam(':is_published', $this->is_published);

        if ($stmt->execute()) {
            $this->id = $this->conn->lastInsertId();
            return true;
        }

        return false;
    }

    public function findById($id) {
        $query = "SELECT c.*, 
                         u.name as instructor_name,
                         cat.name as category_name,
                         (SELECT COUNT(*) FROM lessons WHERE course_id = c.id) as lessons_count,
                         (SELECT COUNT(*) FROM enrollments WHERE course_id = c.id) as students_count,
                         (SELECT AVG(rating) FROM reviews WHERE course_id = c.id) as avg_rating
                  FROM " . $this->table . " c
                  LEFT JOIN users u ON c.instructor_id = u.id
                  LEFT JOIN categories cat ON c.category_id = cat.id
                  WHERE c.id = :id LIMIT 1";

        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(':id', $id);
        $stmt->execute();

        return $stmt->fetch();
    }

    public function getAll($filters = []) {
        $page = $filters['page'] ?? 1;
        $perPage = min($filters['per_page'] ?? DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE);
        $offset = ($page - 1) * $perPage;

        $whereConditions = ['c.is_published = 1'];
        $params = [];

        if (!empty($filters['category_id'])) {
            $whereConditions[] = 'c.category_id = :category_id';
            $params[':category_id'] = $filters['category_id'];
        }

        if (!empty($filters['level'])) {
            $whereConditions[] = 'c.level = :level';
            $params[':level'] = $filters['level'];
        }

        if (!empty($filters['search'])) {
            $whereConditions[] = '(c.title LIKE :search OR c.description LIKE :search)';
            $params[':search'] = '%' . $filters['search'] . '%';
        }

        if (isset($filters['is_featured']) && $filters['is_featured']) {
            $whereConditions[] = 'c.is_featured = 1';
        }

        if (!empty($filters['min_price'])) {
            $whereConditions[] = 'COALESCE(c.discount_price, c.price) >= :min_price';
            $params[':min_price'] = $filters['min_price'];
        }

        if (!empty($filters['max_price'])) {
            $whereConditions[] = 'COALESCE(c.discount_price, c.price) <= :max_price';
            $params[':max_price'] = $filters['max_price'];
        }

        $whereClause = implode(' AND ', $whereConditions);

        // Sorting
        $orderBy = 'c.created_at DESC';
        if (!empty($filters['sort'])) {
            switch ($filters['sort']) {
                case 'price_asc':
                    $orderBy = 'COALESCE(c.discount_price, c.price) ASC';
                    break;
                case 'price_desc':
                    $orderBy = 'COALESCE(c.discount_price, c.price) DESC';
                    break;
                case 'rating':
                    $orderBy = 'avg_rating DESC';
                    break;
                case 'popular':
                    $orderBy = 'students_count DESC';
                    break;
                case 'newest':
                    $orderBy = 'c.created_at DESC';
                    break;
            }
        }

        // Get total count
        $countQuery = "SELECT COUNT(*) as total FROM " . $this->table . " c WHERE $whereClause";
        $countStmt = $this->conn->prepare($countQuery);
        $countStmt->execute($params);
        $total = $countStmt->fetch()['total'];

        // Get paginated results
        $query = "SELECT c.*, 
                         u.name as instructor_name,
                         cat.name as category_name,
                         (SELECT COUNT(*) FROM lessons WHERE course_id = c.id) as lessons_count,
                         (SELECT COUNT(*) FROM enrollments WHERE course_id = c.id) as students_count,
                         (SELECT AVG(rating) FROM reviews WHERE course_id = c.id) as avg_rating
                  FROM " . $this->table . " c
                  LEFT JOIN users u ON c.instructor_id = u.id
                  LEFT JOIN categories cat ON c.category_id = cat.id
                  WHERE $whereClause
                  ORDER BY $orderBy
                  LIMIT :limit OFFSET :offset";

        $stmt = $this->conn->prepare($query);
        foreach ($params as $key => $value) {
            $stmt->bindValue($key, $value);
        }
        $stmt->bindValue(':limit', $perPage, PDO::PARAM_INT);
        $stmt->bindValue(':offset', $offset, PDO::PARAM_INT);
        $stmt->execute();

        return [
            'data' => $stmt->fetchAll(),
            'total' => $total,
            'page' => $page,
            'per_page' => $perPage
        ];
    }

    public function update($id, $data) {
        $fields = [];
        $params = [':id' => $id];

        $allowedFields = ['title', 'description', 'category_id', 'thumbnail', 'price', 'discount_price', 'level', 'duration', 'is_featured', 'is_published'];

        foreach ($data as $key => $value) {
            if (in_array($key, $allowedFields)) {
                $fields[] = "$key = :$key";
                $params[":$key"] = $value;
            }
        }

        if (empty($fields)) {
            return false;
        }

        $query = "UPDATE " . $this->table . " SET " . implode(', ', $fields) . ", updated_at = NOW() WHERE id = :id";
        $stmt = $this->conn->prepare($query);

        return $stmt->execute($params);
    }

    public function delete($id) {
        $query = "DELETE FROM " . $this->table . " WHERE id = :id";
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(':id', $id);

        return $stmt->execute();
    }

    public function getLessons($courseId) {
        $query = "SELECT * FROM lessons WHERE course_id = :course_id ORDER BY order_num ASC";
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(':course_id', $courseId);
        $stmt->execute();

        return $stmt->fetchAll();
    }

    public function getReviews($courseId, $page = 1, $perPage = 10) {
        $offset = ($page - 1) * $perPage;

        $query = "SELECT r.*, u.name as user_name, u.avatar as user_avatar
                  FROM reviews r
                  INNER JOIN users u ON r.user_id = u.id
                  WHERE r.course_id = :course_id
                  ORDER BY r.created_at DESC
                  LIMIT :limit OFFSET :offset";

        $stmt = $this->conn->prepare($query);
        $stmt->bindValue(':course_id', $courseId);
        $stmt->bindValue(':limit', $perPage, PDO::PARAM_INT);
        $stmt->bindValue(':offset', $offset, PDO::PARAM_INT);
        $stmt->execute();

        return $stmt->fetchAll();
    }

    public function getFeatured($limit = 6) {
        $query = "SELECT c.*, 
                         u.name as instructor_name,
                         (SELECT AVG(rating) FROM reviews WHERE course_id = c.id) as avg_rating
                  FROM " . $this->table . " c
                  LEFT JOIN users u ON c.instructor_id = u.id
                  WHERE c.is_featured = 1 AND c.is_published = 1
                  ORDER BY c.created_at DESC
                  LIMIT :limit";

        $stmt = $this->conn->prepare($query);
        $stmt->bindValue(':limit', $limit, PDO::PARAM_INT);
        $stmt->execute();

        return $stmt->fetchAll();
    }

    public function getByInstructor($instructorId) {
        $query = "SELECT c.*,
                         (SELECT COUNT(*) FROM enrollments WHERE course_id = c.id) as students_count,
                         (SELECT AVG(rating) FROM reviews WHERE course_id = c.id) as avg_rating
                  FROM " . $this->table . " c
                  WHERE c.instructor_id = :instructor_id
                  ORDER BY c.created_at DESC";

        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(':instructor_id', $instructorId);
        $stmt->execute();

        return $stmt->fetchAll();
    }
}

File: unified-code-hub-main/src/project-files/flutter/lib/models/course.dart
class Course {
  final int? id;
  final String title;
  final String? description;
  final int? instructorId;
  final String? instructorName;
  final double price;
  final double? discountPrice;
  final double rating;
  final int studentsCount;
  final String? duration;
  final String? category;
  final String level;
  final String? language;
  final String? imageUrl;
  final String? previewVideoUrl;
  final bool isPublished;
  final bool isFeatured;
  final List<Lesson> lessons;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Course({
    this.id,
    required this.title,
    this.description,
    this.instructorId,
    this.instructorName,
    this.price = 0,
    this.discountPrice,
    this.rating = 0,
    this.studentsCount = 0,
    this.duration,
    this.category,
    this.level = 'beginner',
    this.language = 'ar',
    this.imageUrl,
    this.previewVideoUrl,
    this.isPublished = false,
    this.isFeatured = false,
    this.lessons = const [],
    this.createdAt,
    this.updatedAt,
  });

  double get effectivePrice => discountPrice ?? price;

  bool get hasDiscount => discountPrice != null && discountPrice! < price;

  double get discountPercentage {
    if (!hasDiscount) return 0;
    return ((price - discountPrice!) / price * 100);
  }

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'] as int?,
      title: json['title'] as String? ?? '',
      description: json['description'] as String?,
      instructorId: json['instructor_id'] as int?,
      instructorName: json['instructor_name'] as String?,
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      discountPrice: json['discount_price'] != null
          ? (json['discount_price'] as num).toDouble()
          : null,
      rating: (json['rating'] as num?)?.toDouble() ?? 0.0,
      studentsCount: json['students_count'] as int? ?? 0,
      duration: json['duration'] as String?,
      category: json['category'] as String?,
      level: json['level'] as String? ?? 'beginner',
      language: json['language'] as String? ?? 'ar',
      imageUrl: json['image_url'] as String?,
      previewVideoUrl: json['preview_video_url'] as String?,
      isPublished: (json['is_published'] as int?) == 1,
      isFeatured: (json['is_featured'] as int?) == 1,
      lessons: json['lessons'] != null
          ? (json['lessons'] as List).map((l) => Lesson.fromJson(l)).toList()
          : [],
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'instructor_id': instructorId,
      'instructor_name': instructorName,
      'price': price,
      'discount_price': discountPrice,
      'rating': rating,
      'students_count': studentsCount,
      'duration': duration,
      'category': category,
      'level': level,
      'language': language,
      'image_url': imageUrl,
      'preview_video_url': previewVideoUrl,
      'is_published': isPublished ? 1 : 0,
      'is_featured': isFeatured ? 1 : 0,
      'lessons': lessons.map((l) => l.toJson()).toList(),
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}

class Lesson {
  final int? id;
  final int courseId;
  final String title;
  final String? description;
  final String? videoUrl;
  final String? duration;
  final int orderIndex;
  final bool isPreview;
  final String? resources;
  final DateTime? createdAt;

  Lesson({
    this.id,
    required this.courseId,
    required this.title,
    this.description,
    this.videoUrl,
    this.duration,
    this.orderIndex = 0,
    this.isPreview = false,
    this.resources,
    this.createdAt,
  });

  factory Lesson.fromJson(Map<String, dynamic> json) {
    return Lesson(
      id: json['id'] as int?,
      courseId: json['course_id'] as int? ?? 0,
      title: json['title'] as String? ?? '',
      description: json['description'] as String?,
      videoUrl: json['video_url'] as String?,
      duration: json['duration'] as String?,
      orderIndex: json['order_index'] as int? ?? 0,
      isPreview: (json['is_preview'] as int?) == 1,
      resources: json['resources'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'course_id': courseId,
      'title': title,
      'description': description,
      'video_url': videoUrl,
      'duration': duration,
      'order_index': orderIndex,
      'is_preview': isPreview ? 1 : 0,
      'resources': resources,
      'created_at': createdAt?.toIso8601String(),
    };
  }
}
